### ISSUE TITLE
Enhance version catalogs' public API to support tooling

### ISSUE THREAD

**Issue by knisht** [gradle/gradle#20879]
Hello!

### Context

We at JetBrains are interested in supporting version catalogs in a more helpful way than it is done now. For this to be done, we need your help in exposing some implementation details in the Gradle API.

Particularly, we'd like to get more information about version catalogs' structure after the project configuration. 

### Expected Behavior

It would be really helpful to get some sort of mapping from the version catalog's name to the toml file it uses. Certainly, there are other ways to specify version catalog structure (like using DSL in `settings.gradle`), but it looks like toml files are a much more popular format.

### Current Behavior

There is no such API.

**Comment by h0tk3y** [comment]
@knisht 
What are the use cases that you have in mind? What API would you need for that?
When a TOML file is imported, some of its entries might be overridden in scripts. How would you handle that?

**Comment by knisht** [comment]
@h0tk3y Hello!
The main issue I am trying to solve is https://youtrack.jetbrains.com/issue/IDEA-279603.
I expect some kind of a listener which is invoked when the version catalogs machinery accesses some `*.toml` file. It also may be some class that provides a `Map` from version catalog name to a `*.toml`.
Regarding overriding in scripts, this is a problem indeed. I cannot think of a way how to solve it reliably, but it doesn't mean that we should decline this feature.

**Comment by melix** [comment]
A TOML fle can be imported from a local file or a remote one (e.g https://melix.github.io/blog/2022/02/micronaut-version-catalog.html). It's only used to generate accessors. If the file didn't change, it won't even be read, so "accessing the TOML file" isn't the right thing either. The TOML file is just _one_ source for the catalog.

There is a similar need for automating dependency upgrades (see what we do for Micronaut: https://github.com/micronaut-projects/micronaut-build/blob/ec7d227d1e1b263b94f1b2ecc8ac91f36bd63225/src/main/java/io/micronaut/build/catalogs/MicronautVersionCatalogUpdatePlugin.java#L17).

Basically there are I think a couple requirements:

1. accessing the "model" of a catalog, before it's actually realized
2. adding some "source" to the model, so that tooling can do things like link to the source files

As a side note, the generated accessors contain some comments linking to the source catalog file, but it's a poor man's solution since it doesn't quite make sense for external catalogs.

**Comment by nskvortsov** [comment]
IDEA needs an access to a model of version catalogues to provide better navigation. For now, it is only possible to navigate to generated accessors. Which is useless for most of our users. They expect IDEA to go to alias definitions: in settings.gradle or toml files. Former is hard to handle (Groovy/KTS), but latter syntax is not dynamic and we can start with it.

related ticket  https://youtrack.jetbrains.com/issue/IDEA-282725/Navigate-from-versionCatalogs-alias-to-settingsgradle
